editor/editor_tool_drawer.cpp

```diff
@@ -0,0 +1,496 @@
+	tb_label->set_v_grow_direction(GROW_DIRECTION_BOTH);
+	tb_label->set_vertical_alignment(VERTICAL_ALIGNMENT_CENTER);
+	tb_label->set_text(p_name);
+	tb->add_child(tb_label);
+
+	// Create the drawer section for the drawer control.
+	EditorToolDrawerSection *section = memnew(EditorToolDrawerSection);
+	section->set_title(p_name);
+	section->set_title_icon(p_icon);
+	section->hide();
+	tool_drawer->add_child(section);
+
+	section->add_content(p_control);
+
+	// Connect input signals.
+	tb->connect(SNAME("pressed"), callable_mp(this, &EditorToolDrawer::_toggle_section), varray(tb, section, p_control));
+
+	// Map the drawer control to the related nodes and data.
+	tool_drawer_button_map[p_control] = tb;
+	tool_drawer_section_map[p_control] = section;
+	tool_drawer_expanded_map[p_control] = false;
+
+	bool force_display_drawers = EDITOR_GET("editors/common/always_display_drawers");
+	if (force_display_drawers) {
+		tb->set_pressed(true);
+		section->set_visible(true);
+		tool_drawer_expanded_map[p_control] = true;
+	}
+
+	_update_mass_toggle_visibility();
+	_update_mass_toggle_icon();
+}
+
+void EditorToolDrawer::remove_control_from_drawer(Control *p_control) {
+	ERR_FAIL_COND(!p_control);
+
+	// Remove the activator button.
+	if (tool_drawer_button_map.has(p_control)) {
+		Button *tb = tool_drawer_button_map[p_control];
+		tool_drawer_button_map.erase(p_control);
+		tool_drawer_bar->remove_child(tb);
+		memdelete(tb);
+	}
+
+	// Remove the drawer section.
+	if (tool_drawer_section_map.has(p_control)) {
+		EditorToolDrawerSection *section = tool_drawer_section_map[p_control];
+		tool_drawer_section_map.erase(p_control);
+		tool_drawer_bar->remove_child(section);
+		memdelete(section);
+	}
+
+	// Remove the expanded state mapping.
+	if (tool_drawer_expanded_map.has(p_control)) {
+		tool_drawer_expanded_map.erase(p_control);
+	}
+
+	_update_mass_toggle_visibility();
+	_update_mass_toggle_icon();
+}
+
+void EditorToolDrawer::set_control_visible(Control *p_control, bool p_visible) {
+	ERR_FAIL_COND(!p_control);
+
+	bool expanded = false;
+	bool force_display_drawers = EDITOR_GET("editors/common/always_display_drawers");
+	if (force_display_drawers) {
+		expanded = true;
+	} else if (tool_drawer_expanded_map.has(p_control)) {
+		expanded = tool_drawer_expanded_map[p_control];
+	}
+
+	if (tool_drawer_button_map.has(p_control)) {
+		Button *tb = tool_drawer_button_map[p_control];
+		tb->set_visible(p_visible);
+		if (p_visible) {
+			tb->set_pressed(expanded);
+		} else {
+			tb->set_pressed(false);
+		}
+	}
+
+	if (tool_drawer_section_map.has(p_control)) {
+		EditorToolDrawerSection *section = tool_drawer_section_map[p_control];
+		if (p_visible) {
+			section->set_visible(expanded);
+		} else {
+			section->set_visible(false);
+		}
+	}
+
+	_update_tool_drawer();
+	_update_mass_toggle_visibility();
+	_update_mass_toggle_icon();
+}
+
+void EditorToolDrawer::_toggle_section(Control *p_button, Control *p_section, Control *p_control) {
```

Comment:
`p_button` type can be changed to Button after recent binding changes.

Reply:
Or actually, if you connect the `toggled` signal, you don't even need to pass the button.

Reply:
We don't even need a reference to the section, especially now that we have a single map with a struct. I don't remember what was my rationale.

Reply:
I had to leave it using `pressed` because of how `toggled` is triggered on `set_pressed` calls, which makes the logic very messy. Doesn't matter for the method signature, since I can still use the unified map and the struct to get the current state.